File: lib/bubble_sort.c
#include "libc.h"

void bubble_sort(int *arr, int size)
{
   int max = size - 1;
   int i;
   while (max > 1)
   {
      for (i = 0; i < max; ++i)
      {
         if (arr[i] > arr[i+1])
            swap(&arr[i], &arr[i+1]);
            PRINT(bubble, arr, size);
      }
      --max;
   }
}

File: rust/algorithms/src/bubblesort.rs

use crate::defs::*;


pub fn bubble_sort<T: PartialOrd + Clone>(arr: &mut [T], order: ORDER) {
    //type CMP<T> = fn (i: &T, j : &T) -> bool;
    let cmp: CMP<T>;
    match order {
        ORDER::ASCENT => {
            cmp = is_first_one_larger;
        }
        ORDER::DESCENT => {
            cmp = is_first_one_smaller;
        }
    }
    for i in 0..arr.len() {
        for j in 0..(arr.len()-1-i) {
            if cmp(&arr[j], &arr[j+1]) {
                swap(arr, j, j+1);
            }
        }
    }
}
